### ISSUE TITLE
core/generator: data race in getAndAddBlockSignatures

### ISSUE THREAD

**Issue by jbowens** [chain/chain#988]
Running `go test -race` in chain/core/generator reveals a data race in the TestGetAndAddBlockSignatures test.

```
==================
WARNING: DATA RACE
Read at 0x00c4201b4558 by goroutine 29:
  chain/protocol/bc/legacy.mapBlockHeader()
      /Users/jackson/src/chain/protocol/bc/legacy/map.go:296 +0x3ab
  chain/protocol/bc/legacy.(*BlockHeader).Hash()
      /Users/jackson/src/chain/protocol/bc/legacy/block_header.go:67 +0x3c
  chain/core/generator.testSigner.SignBlock()
      /Users/jackson/src/chain/core/generator/generator_test.go:115 +0x52
  chain/core/generator.(*testSigner).SignBlock()
      <autogenerated>:2 +0xc7
  chain/core/generator.getSig()
      /Users/jackson/src/chain/core/generator/block.go:141 +0x80

Previous write at 0x00c4201b4558 by goroutine 27:
  chain/core/generator.(*Generator).getAndAddBlockSignatures()
      /Users/jackson/src/chain/core/generator/block.go:126 +0x115a
  chain/core/generator.TestGetAndAddBlockSignatures()
      /Users/jackson/src/chain/core/generator/generator_test.go:80 +0x5fc
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:657 +0x107

Goroutine 29 (running) created at:
  chain/core/generator.(*Generator).getAndAddBlockSignatures()
      /Users/jackson/src/chain/core/generator/block.go:105 +0x38b
  chain/core/generator.TestGetAndAddBlockSignatures()
      /Users/jackson/src/chain/core/generator/generator_test.go:80 +0x5fc
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:657 +0x107

Goroutine 27 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:697 +0x543
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:882 +0xaa
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:657 +0x107
  testing.runTests()
      /usr/local/go/src/testing/testing.go:888 +0x4e0
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:822 +0x1c3
  main.main()
      chain/core/generator/_test/_testmain.go:46 +0x20f
==================
--- FAIL: TestGetAndAddBlockSignatures (0.00s)
	testing.go:610: race detected during execution of test
```

**Comment by jbowens** [comment]
I think this is a real (albeit unlikely) race when configured for _m_ of _n_ block signatures. When we call getAndAddBlockSignatures, we launch a goroutine for each signer. That goroutine reads the newly-created block so that it can send the marshaled bytes to the remote signer or to calculate its hash if it's the local mock HSM. Once we get enough good signatures, we write the signatures to that same block struct. It's _possible_ that one of the get-signature goroutines is really slow still hasn't read the block.

There's a TODO in cmd/cored:
```
	// TODO(kr): We might end up serializing b multiple
	// times in multiple calls to different remoteSigners.
	// Maybe optimize that if it makes a difference.
```
Marshaling once before launching the remote calls would also solve this issue. Any objections to doing that? If the local mock HSM is one of the signers, this will be an unnecessary serialization and deserialization, but we probably shouldn't optimize for the mock HSM anyways.

**Comment by jbowens** [comment]
#1046 inadvertently caused the TestGetAndAddBlockSignatures test to no longer exercise the race, but I'm confident it's still there. I'll try to add a new test in #1033 that exercises the race.

**Comment by kr** [comment]
> Marshaling once before launching the remote calls would also solve this issue. Any objections to doing that? If the local mock HSM is one of the signers, this will be an unnecessary serialization and deserialization, but we probably shouldn't optimize for the mock HSM anyways.

Yeah good point. MockHSM performance is prob not important enough to justify making the generator's code more complicated.

**Comment by jbowens** [comment]
Fixed in #1033.
